actions/rules/action_state_to_metastate.go

```diff
@@ -0,0 +1,416 @@
+	}
+	return wit, nil
+}
+
+func (act ActionStateToMetaState) loadWorkItemByID(id uuid.UUID) (*workitem.WorkItem, error) {
+	if act.Ctx == nil {
+		return nil, errors.New("Context is nil")
+	}
+	if act.Db == nil {
+		return nil, errors.New("Database is nil")
+	}
+	wi, err := act.Db.WorkItems().LoadByID(act.Ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	return wi, nil
+}
+
+func (act ActionStateToMetaState) storeWorkItem(workitem *workitem.WorkItem) (*workitem.WorkItem, error) {
+	if act.Ctx == nil {
+		return nil, errors.New("Context is nil")
+	}
+	if act.Db == nil {
+		return nil, errors.New("Database is nil")
+	}
+	if act.UserID == nil {
+		return nil, errors.New("UserID is nil")
+	}
+	err := application.Transactional(act.Db, func(appl application.Application) error {
+		var err error
+		workitem, err = appl.WorkItems().Save(act.Ctx, workitem.SpaceID, *workitem, *act.UserID)
+		if err != nil {
+			return errors.New("Error updating work item")
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return workitem, nil
+}
+
+func (act ActionStateToMetaState) getValueListFromFieldType(wit *workitem.WorkItemType, fieldName string) ([]interface{}, error) {
+	fieldType := wit.Fields[fieldName].Type
+	switch t := fieldType.(type) {
+	case workitem.EnumType:
+		return t.Values, nil
+	}
+	return nil, errors.New("Given field on workitemtype " + wit.ID.String() + " is not an enum field: " + fieldName)
+}
+
+func (act ActionStateToMetaState) getStateToMetastateMap(workitemTypeID uuid.UUID) (map[string]string, error) {
+	wit, err := act.loadWorkItemTypeByID(workitemTypeID)
+	if err != nil {
+		return nil, err
+	}
+	stateList, err := act.getValueListFromFieldType(wit, workitem.SystemState)
+	if err != nil {
+		return nil, err
+	}
+	metastateList, err := act.getValueListFromFieldType(wit, workitem.SystemMetaState)
+	if err != nil {
+		return nil, err
+	}
+	stateToMetastateMap := make(map[string]string)
+	for idx := range stateList {
+		thisState, ok := stateList[idx].(string)
+		if !ok {
+			return nil, errors.New("State value in value list is not of type string")
+		}
+		thisMetastate, ok := metastateList[idx].(string)
+		if !ok {
+			return nil, errors.New("Metastate value in value list is not of type string")
+		}
+		stateToMetastateMap[thisState] = thisMetastate
+	}
+	return stateToMetastateMap, nil
+}
+
+func (act ActionStateToMetaState) getMetastateToStateMap(workitemTypeID uuid.UUID) (map[string]string, error) {
+	stateToMetastate, err := act.getStateToMetastateMap(workitemTypeID)
+	if err != nil {
+		return nil, err
+	}
+	metastateToStateMap := make(map[string]string)
+	for state, metastate := range stateToMetastate {
+		metastateToStateMap[metastate] = state
+	}
+	return metastateToStateMap, nil
+}
+
+func (act ActionStateToMetaState) addOrUpdateChange(changes *[]convert.Change, attributeName string, oldValue interface{}, newValue interface{}) []convert.Change {
```

Comment:
Consider making a `type ChangeSet []convert.Change` and putting such members as methods onto that new type.

Reply:
Next PR.